testsuite/integration-arquillian/tests/other/base-ui/src/test/java/org/keycloak/testsuite/ui/account2/page/ApplicationsPage.java

```diff
@@ -17,14 +17,94 @@
 
 package org.keycloak.testsuite.ui.account2.page;
 
+import org.keycloak.testsuite.util.UIUtils;
+import org.openqa.selenium.By;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * @author Vaclav Muzikar <vmuzikar@redhat.com>
  */
 public class ApplicationsPage extends AbstractLoggedInPage {
+    @FindBy(xpath = "//li[starts-with(@id,'application-client-id')]")
+    private List<WebElement> applications;
+
     @Override
     public String getPageId() {
         return "applications";
     }
 
-    // TODO implement this! (KEYCLOAK-12107)
+    public List<ClientRepresentation> toggleApplicationsDetails() {
+        ArrayList<ClientRepresentation> apps = new ArrayList<>();
+        for(WebElement app : applications) {
+            String clientId = app.getAttribute("id").replace("application-client-id-", "");
+            WebElement expandButton = app.findElement(By.xpath("//button[@id='application-toggle-" + clientId + "']"));
+            expandButton.click();
+            apps.add(toRepresentation(app, clientId));
+        }
+        return apps;
+    }
+
+    public List<ClientRepresentation> getApplications() {
+        ArrayList<ClientRepresentation> apps = new ArrayList<>();
+        for(WebElement app : applications) {
+            String clientId = app.getAttribute("id").replace("application-client-id-", "");
+            apps.add(toRepresentation(app, clientId));
+        }
+        return apps;
+    }
+
+    private ClientRepresentation toRepresentation(WebElement app, String clientId) {
+        String clientName = UIUtils.getTextFromElement(app.findElement(By.xpath("//div[@id='application-name-" + clientId + "']")));
+        boolean userConsentRequired = !UIUtils.getTextFromElement(app.findElement(By.xpath("//div[@id='application-internal-" + clientId + "']"))).equals("Internal");
+        boolean inUse = UIUtils.getTextFromElement(app.findElement(By.xpath("//div[@id='application-status-" + clientId + "']"))).equals("In use");
+        String baseURL = UIUtils.getTextFromElement(app.findElement(By.xpath("//div[@id='application-baseurl-" + clientId + "']")));
+        boolean applicationDetailsVisible = app.findElement(By.xpath("//section[@id='expandable']")).isDisplayed();
```

Comment:
I'm not sure how this can work. The ID doesn't have a unique value at all so we cannot be sure what app was selected. This is of course not an issue with the test as such but rather with the page.